feat(specs,tests): Implement EIP-7928 Block-Level Access Lists

fselmo opened the pull request:
## 🗒️ Description

Implement EIP-7928: Block-Level Access Lists

## 🔗 Related Issues or PRs

Ported from [fselmo/execution-specs branch](https://github.com/fselmo/execution-specs/tree/feat/amsterdam-fork-and-block-access-lists)

## ✅ Checklist
<!-- Please check off all required items. For those that don't apply remove them accordingly. -->

- [x] All: Ran fast `tox` checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    uvx tox -e static
    ```
- [x] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.
- [x] All: Considered adding an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [x] All: Set appropriate labels for the changes (only maintainers can apply labels).
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://eest.ethereum.org/main/tests/) are correctly formatted.
- [x] Tests: For PRs implementing a missed test case, update the [post-mortem document](/ethereum/execution-spec-tests/blob/main/docs/writing_tests/post_mortems.md) to add an entry the list.
- [x] Ported Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) or [tests/static](/ethereum/execution-spec-tests/blob/main/tests/static) have been assigned `@ported_from` marker.

#### Cute Animal Picture

<img width="1249" height="876" alt="Screenshot 2026-01-16 at 10 11 17" src="https://github.com/user-attachments/assets/3ffbb155-5278-4b45-a923-348b3134eb9a" />

raxhvl committed:
feat(tests): Implement more EIP-7928 tests

refactor(tests): Prevent skips by splitting tests appropriately

fix(tests): Use valid inputs to precompile tests

chore(tests): linting fixes

feat(tests): EIP-7928 test_bal_storage_write_read_cross_frame

feat(tests): EIP-7928 test_bal_storage_write_read_same_frame

feat(tests): EIP-7928 test_bal_nonexistent_account_access

feat(tests): EIP-7928 test_bal_nonexistent_value_transfer

feat(tests): EIP-7928 test_bal_precompiles

feat(tests): EIP-7928 test_bal_withdrawal_to_coinbase_empty_block

feat(tests): EIP-7928 test_bal_withdrawal_to_coinbase

feat(tests): EIP-7928 test_bal_withdrawal_largest_amount

feat(tests): EIP-7928 test_bal_withdrawal_to_precompiles

fix(tests): expectation for nonexistent account in post

fix(specs,tests): Fix withdrawal tests for BALs issue with idx==0

- `self.txs.successfully_parsed` is a list of transaction indexes, not
  transactions. The "if tx" check here would then check `if 0` which
  parses as a boolean ``False``. This means we would skip counting the tx
  if index=0 was successful.

- Fixes some test expectations where `post_code` was being checked instead
  of ``new_code``.

feat(tests): EIP-7928 test_bal_zero_withdrawal

feat(tests): EIP-7928 test_bal_withdrawal_and_new_contract

feat(tests): EIP-7928 test_bal_withdrawal_and_selfdestruct

feat(tests): EIP-7928 test_bal_multiple_withdrawals_same_address

feat(tests): EIP-7928 withdrawal_and_value_transfer_same_address

feat(tests): EIP-7928 withdrawal_and_state_access_same_account

feat(tests): EIP-7928 test_bal_withdrawal_no_evm_execution

feat(tests): EIP-7928 test_bal_withdrawal_to_nonexistent_account

feat(tests): EIP-7928 test_bal_withdrawal_empty_block

feat(tests): EIP-7928 test_bal_withdrawal_with_transaction

feat(tests): EIP-7928 coinbase

fselmo committed:
fix(specs): Fix zero value withdrawals BAL tracking (#29)

* fix(specs): Fix zero value withdrawals BAL tracking

* docs(specs): rename 'finalize' to 'normalize' in comments

* docs(specs): remove reference to uint128 for balance tracking

---------

Co-authored-by: Toni Wahrstätter <51536394+nerolation@users.noreply.github.com>
Co-authored-by: Toni Wahrstätter <info@toniwahrstaetter.com>

nerolation committed:
fix(specs): static upfront check for create + selfdestruct (#22)

nerolation committed:
fix(specs): Ensure tracking before first access (#1722)

fselmo committed:
fix(spec-specs): duplicate storage writes in state tracker (#1743)

- Perform a similar check to balance changes and other
  tracker methods and keep only the last write.

nerolation committed:
feat(specs): EIP-7928 refactoring

fselmo committed:
fix(spec-specs): require and use blockenv for state tracking

fselmo committed:
refactor(spec-specs): track BAL changes via frames

fselmo committed:
feat(specs): Implement EIP-7928: Block-Level Access Lists

fix(tests): Fix Amsterdam filling after rebase
fix(specs): Fix issues with new ruff + mypy rules after rebase

- bal -> block_access_list; re-add custom rlp encoding for block access list
- bytes to uint
- move away from method-style

- Update EIP-7928 implementation: system contracts at index 0, migrate to RLP
  - System contracts (parent hash, beacon root) now use block_access_index 0
  - Transactions use block_access_index 1 to len(transactions)
  - Post-execution changes use block_access_index len(transactions) + 1
  - Migrated from SSZ to RLP encoding as per updated EIP-7928 spec
  - Updated all tests to match new API and structure
  - Replaced tx_index with block_access_index throughout codebase

- add system contract logic
- add markdown docstrings
- update BAL format; address comments
- ssz encoding and bal validation
- six ssz types
- bal tests
- balspecs

fix: do not track setting empty code to a new account (#19)
fix: track implicit SLOAD within SSTORE for OOG cases (#18)
refactor: Put back explicit acct tracking outside 7702 delegation path (#17)
fix non-tracked 7702 authority for invalid delegations (#16)
    * fix non-tracked 7702 authority for invalid delegations
    * fix: lint issues
    * fix: track delegation target when loaded as call target
    * fix: track delegation target when loaded as call target from call
      opcodes
    * chore: fix issues with documentation generation

Fix self-destruct cases with pre-execution balance cache / tracking
    * fix self-destruct implementation
    * fix self-destruct tracking balance
    * fix it in the bal finalization by filtering
    * add balance reset and fix tests
    * simplify pre-balance tracking not using snapshots

fix duplicated code entries for in transaction self destruct
fix self destruct in same transaction bug
fix call/delagate call tracking bug
fix zero-value transfer tracking (#6)
    * fix zero-value transfer tracking
    * fix reverted frame tracking
    * rename variables
    * fix missing addresses bug
    * fix: docs run & move imports to top of file

refactor: move rlp_utils to block_access_lists; bal -> block_access_lists
Some remaining fixes due to large refactor in `forks/osaka`:

- Move BALs from amsterdam -> forks/amsterdam
- rename: build -> build_block_access_list
- fix docc issues

move state change tracker to State

correct system contract addresses

Fixes to communicate with BALs EEST branch:

- fix(bal): Initialize the state tracker before system contract calls

- We were missing system contract calls to beacon roots and history
  contracts. This change initializes the state tracker before system
  contract calls and passes the tracker to these calls if post-Amsterdam.

- fix(docs): Fix issues with toxenvs: lint, doc, json_infra
- fix(t8n): Only initialize the bal_change_tracker for amsterdam
- feat(fork criteria): Index upcoming forks for better ordering / fix issues
- chore(forks): Fix issues from lint after rebase with Osaka latest
- fix(setuptools): Update packages to include amsterdam
- chore(lint): Fix 'tox -e static' issues

- Fix bug in tracker
  Manually cherry-picked from e72991bf3876563900d5c2bcc2442b0a1eeb439f
  Author: nerolation

- chore(tests): Attempt to resolve issues with CI tests
- chore(lint): fix issues from running ``tox -e static`` locally
- refactor(bal): Send BAL as a list over t8n tool
- fix(amsterdam): Add change tracker to state test in t8n
- chore(lint,tests): Fix tests after moving bal from osaka -> amsterdam
- chore(forks): Move bals from Osaka to Amsterdam
- chore(lint): Fix lint issues

- refactor(bal): Send the full bal object and bal_hash over t8n

  - If we send the full object over JSON, we can model_validate() on ESST.
  - If we send the hash, once we fill the pydantic model, we can get the rlp
    and the hash and validate that our objects match while only really
    validating the parts of the BAL we are interested in for each test.

- chore: point to working eest branch
- chore(bals): Remove unused SSZ utils.py
  The SSZ implementation is no longer needed as we are now using RLP
- refactor(bals): Clean up BAL module types and imports
  - Bytes -> Bytes32 type for storage slots
  - Remove unused imports / fix imports / fix linting
  - Update function signatures to match tracker

- fix(bals-tx-index): Track bal indexes in t8n
  Keep track of BAL index state in t8n

fselmo committed:
fix(spec-specs): Mark original addr warm before delegation

Co-authored-by: spencer spencer.taylor-brown@ethereum.org

fselmo committed:
fix(spec-specs): Make sure we account for no changes

fselmo committed:
fix(spec-specs): Better tracking for code changes; ensure with BAL test

fselmo committed:
fix(spec-specs): Use child frame for create message

fselmo committed:
fix(spec-specs): Normalize transaction before merging to block frame

fselmo committed:
fix(spec-specs): Early static check for SSTORE before any reads

fselmo committed:
fix(spec-specs): Track storage writes more appropriately wrt index

fselmo committed:
fix(test-specs): validate t8n BAL independent of expectation existence (#1742)

- Validate static checks on the t8n BAL if it exists
- IF the expectation also exists, validate against the expectation

Keep these checks separate as this helps validation now that we fill
for all tests, regardless if they have an expectation or not.

fselmo committed:
fix(spec-specs): Use functions, not methods; fix create revert

fselmo committed:
fix(spec-specs): Default code to b"" in tracker, skip empty setting

fselmo committed:
fix(spec-specs): Fix BAL cross-transaction tracking and nonce dedup

 - add commit_transaction_frame() - no net-zero filtering
   for cross-tx changes
 - keep max nonce per transaction when building BAL, remove
   block-level code filtering
 - filter net-zero code changes at tracking time
   (for 7702 txs)
 - use commit_transaction_frame() instead of
   merge_on_success() for tx->block commits

fselmo committed:
fix(spec-specs): Move destroy_account before BAL normalization

fselmo committed:
fix(spec-specs): Check delegation access gas before reading

fselmo committed:
fix(spec-specs): Track code per auth; filter pre at tx frame

fselmo committed:
fix(spec-specs): Use proper frames for system transactions

fselmo committed:
fix(spec-specs): Track address at init collision

fselmo committed:
chore(spec-specs): Add Amsterdam docstring; update prepare msg

fselmo committed:
chore: Add pre-amsterdam BAL tests to doc for tracking

fselmo committed:
fix(spec-specs): Calculate all gas we can before accessing state

- Calculate all gas that we can without state access and check this
  gas before ever accessing state. This is the most sensible way for
  an implementation to behave and indeed was revealed to be the way
  clients are behaving, which differed from the specs.

- Use fork.gas_costs to calculate gas costs, NOT hard-coded values.

- Create a BAL expectation for the test that yielded discrepancies
  between clients and specs so this doesn't slip through again.
  Document this test in `test_cases.md`.

fselmo committed:
fix(test-tools): Remove named forks from blobSchedule; turn off BPOs

felix314159 committed:
fix(cli): add bal exception for erigon (#1809)

fselmo committed:
feat(test): Better describe the BAL for selfdestruct revert

qu0b committed:
test(eip7928): add EXTCODECOPY OOG memory expansion BAL test

qu0b committed:
test(eip7928): add cross-block precompile state leak test

fselmo committed:
refactor(test-tests): parametrize existing test oog case instead

fselmo committed:
refactor(test-tests): Add BAL expectation to state leak test; fix lint

fselmo committed:
feat(tests): Port oog create refund test; add BAL >= Amsterdam

qu0b committed:
feat(test): add SELFDESTRUCT OOG BAL test

fselmo committed:
refactor(tests): move selfdestruct bal tests to oog file; add gas boundaries

- more updates to test_cases.md

fselmo committed:
refactor(spec-specs): Refactor state changes and frame hierarchy (#1841)

* refactor(spec-specs): Refactor state changes and their frames

* chore(spec-specs): cleanup BAL logic; organize gas check for SSTORE

* refactor(spec-specs): Changes from comments on PR #1841

* enhance: don't set defaults for state_changes

gurukamath committed:
fix(spec,tests): Change BAL to List[AccountChange] (#1844)

* fix(spec): update BAL type

* fix(tests): run amsterdam jsons

* fix(test-specs): Add BAL to genesis beyond Amsterdam fork

---------

Co-authored-by: fselmo <fselmo2@gmail.com>

fselmo committed:
fix(spec-specs): Fix issues after rebasing with forks/osaka

- fix mkdocs

qu0b committed:
feat(test): add more bal test cases (#1812)

* rebase onto upstream
* chore(fix) balance check in call before target access
* merge test cases with usptream
* chore(fix) format with ruff
* chore(fix) revert call changes and add target to bal
* merge test cases with usptream
* improve wording
* chore(formate) fix formatting and line length
* refactor(test-tests): Use pre API where possible; explicit check for none in BAL
* refactor(test-tests): Refactor opcode tests to bal opcodes test file

- Qu0b/fix pre alloc group enginex (#1911)

---------

Co-authored-by: fselmo <fselmo2@gmail.com>

fselmo committed:
feat(test-tests): BAL test for nested storage write reset same tx (#1854)

fselmo committed:
fix(spec-specs): post-exec net-zero filtering post specs refactor

- refactor(spec-specs): Move net-zero filtering inside commit tx frame

fselmo committed:
feat(test-tests): Expand BAL CALL opcode OOG boundary test cases (#1882)

* feat(test-tests): expand bal call opcode oog boundary tests

Parametrize for:

- value / no value
- COLD / WARM target
- 7702 delegation / no delegation
  - WARM / COLD delegation
- mem expansion / no mem expansion

Include tests both before and after state access to ensure BAL
expectations at these gas boundaries are met.

* refactor(tests): Use `fork.memory_expansion_gas_calculator()`

* feat(test-tests): Refactor; Add second 7702 boundary at success minus 1

* refactor: changes from comments on PR #1882

* fix: update test names and descriptions after refactor

fselmo committed:
feat(spec): update eip7928 to latest rlp specs wrt storage; rename tx_index (#1912)

* feat(spec-test): Change storage RLP encoding to U256; tx_index rename

* fix: unit tests exception message

fselmo committed:
fix(test-tests): Avoid hard-coding precompile range for lexicographic test

nerolation committed:
feat(specs): EIP-7928 move bal from payload (#1917)

* feat(specs): EIP-7928 move bal from payload

* remove BAL from stf

* fix linter

fselmo committed:
feat(test): OOG and success selfdestruct tests to all precompiles

feat(test): Expand fork range for selfdestruct to precompile tests

- For successful tests, start at Homestead where precompiles were
  introduced (EIPs 196, 197, 198).

- For OOG tests, start at Tangerine where operation gas costs were
  introduced (EIP 150).

fselmo committed:
fix(test-tests): Use `ZeroPaddedHexNumber` instead of `HexNumber` for BALs (#1922)

* fix(test-tests): Use ZeroPaddedHexNumber instead of HexNumber for consistency

* chore(test-tests): add bal serialization roundtrip & 0-padded hex test (#33)

* refactor(test): Remove unnecessary instantiation of classes to pydantic types

---------

Co-authored-by: danceratopz <danceratopz@gmail.com>

fselmo committed:
refactor(spec-specs): Refactor specs to be more coherent wrt gas acco… (#1897)

* refactor(spec-specs): Refactor specs to be more coherent wrt gas accounting

* feat(test): BAL test for call with value in static context

fselmo committed:
refactor(test): `Tangerine` -> `TangerineWhistle`; comments on PR #1954

Bonus:

- fix(test): remove unnecessary isolation for enginex

fselmo committed:
fix(tool): Fix EvmOneTransitionTool parsing for TangerineWhistle (add space)

nerolation committed:
feat(tests): add more 7928 test descriptions (#1815)

* feat(tests): add more 7928 test descriptions

* chore(test): remove test duplicated by test_bal_create_selfdestruct_to_self_with_call

---------

Co-authored-by: fselmo <fselmo2@gmail.com>

fselmo committed:
feat(test): Extend selfdestruct tests to all Amsterdam gas boundaries

raxhvl committed:
feat(tests): EIP-7928 tests targeting EIP-4788 (#1887)

* ✨ feat(tests): 7928x4788 tests

* 🧹 chore: Rename bal index

* ✨ feat: zero indexed tx

* 🥢 nit: Balance change for query contract

* 🥢 nit: Exclude system address from BAL

* 🥢 nit: Exclude system address from BAL

* 📄 docs: Update test description

---------

Co-authored-by: raxhvl <raxhvl@users.noreply.github.com>
Co-authored-by: felipe <fselmo2@gmail.com>

nerolation committed:
feat(tests): add invalid BAL tests for spurious block_access_index (#1953)

raxhvl committed:
feat(tests): EIP-7928 tests targeting EIP-7002 (#1918)

* feat(test-tests): Add EIP-7002 BAL tests

🥢 nit:

✨ feat(tests): parameterised amount in clean sweep

✨ feat(tests): test_bal_7002_request_invalid

✨ feat(tests): test_bal_withdrawal_request_from_contract

✨ feat(tests): test_bal_7002_no_withdrawal_requests

♻️ refactor:

✨ feat(tests): parameter: validator key

✨ feat(tests): simplify

✨ feat(tests): test_bal_7002_partial_sweep

✨ feat(tests): test_bal_7002_clean_sweep

✨ feat: add more coverage

✨ feat:  test_bal_withdrawal_request_from_eoa

* fix(test-tests): lint

* chore: update test docstring to match case description and test behavior

* fix: balance_changes -> balance for post Accounts

---------

Co-authored-by: raxhvl <raxhvl@users.noreply.github.com>
Co-authored-by: fselmo <fselmo2@gmail.com>

raxhvl committed:
feat(tests): Test extraneous entries for BAL (#1992)

* 🧪 test(EIP-7928): test_bal_invalid_extraneous_entries

* 🥢 nit:

* chore: fix lint issues

---------

Co-authored-by: raxhvl <raxhvl@users.noreply.github.com>
Co-authored-by: fselmo <fselmo2@gmail.com>

fselmo committed:
refactor(test): fork.supports_protected_txs() instead of direct comparison

- This will also help with support for any forks of this repo for L2s
  or other projects if some fork definitions don't have any meaningful
  relationship to SpuriousDragon.

fselmo committed:
refactor: address comments from PR #1954

fselmo committed:
feat(test): Add selfdestruct to system contracts + to self from initcode

fselmo committed:
feat(test-commands): Allow pytest valid fork markers as params; add unit tests

fselmo committed:
refactor(spec): Remove `check_gas` where unnecessary for instructions

gurukamath committed:
refactor(eip7928): refactor net zero filtering in BALs (#1899)

* refactor(eip7928): refactor net zero filtering in BALs

* refactor(eip7928): handle selfdestruct correctly

fselmo committed:
fix(test-fill): organize alloc groups by pytest param (enginex)

gurukamath committed:
fix(tests): remove bal from block body (#2017)

* tests(eip-7928): remove bal from block body

* fix(tests): clean up json_infra and genesis bal in block

* fix(tests): fix add_genesis_block empty bal hash

---------

Co-authored-by: fselmo <fselmo2@gmail.com>

fselmo committed:
feat(spec-tests): Update json_infra tests version (#1939)

spencer-tb commented on packages/testing/src/execution_testing/cli/pytest_commands/plugins/consume/simulators/helpers/ruleset.py:
```diff
@@ -296,7 +296,6 @@ def get_blob_schedule_entries(fork: Fork) -> Dict[str, int]:
         "HIVE_CANCUN_TIMESTAMP": 0,
         "HIVE_PRAGUE_TIMESTAMP": 0,
         "HIVE_OSAKA_TIMESTAMP": 0,
-        **get_blob_schedule_entries(Osaka),
```
Why do we remove these? I think this might break some of the existing pre-Amsterdam tests.

fselmo replied:
Ah ty for catching. I thought that named forks got rid of blob schedule entries. I'm still unsure how this should work but this does get tests passing for Amsterdam (I turned Amsterdam back on recently as well) but missed turning back on for Osaka. If you have more insight, can you elaborate on what this _should_ be?

fselmo replied:
It looks like we still need to pass these in hive for some reason for now... so yeah we should keep this 👍🏼

spencer-tb replied:
"I thought that named forks got rid of blob schedule entries." ahh yep! We can consolidate in the near future.

spencer-tb commented on packages/testing/src/execution_testing/test_types/block_access_list/modifiers.py:
```diff
@@ -242,46 +258,62 @@ def transform(bal: BlockAccessList) -> BlockAccessList:
                     for storage_change in storage_slot.slot_changes:
-                        if storage_change.tx_index == tx1:
+                        if storage_change.block_access_index == tx1:
                             balance_indices[tx1] = True
```
Should this be storage here?
```suggestion
                            storage_indices[tx1] = True
```

spencer-tb commented on packages/testing/src/execution_testing/test_types/block_access_list/modifiers.py:
```diff
@@ -242,46 +258,62 @@ def transform(bal: BlockAccessList) -> BlockAccessList:
+                                ZeroPaddedHexNumber(tx2)
+                            )
+                        elif storage_change.block_access_index == tx2:
                             balance_indices[tx2] = True
```
Same as above?
```suggestion
                            storage_indices[tx2] = True
```

spencer-tb commented on src/ethereum/forks/amsterdam/state_tracker.py:
```diff
@@ -0,0 +1,560 @@
+
+    # Pre-state captures (transaction-scoped, only populated at tx frame)
+    pre_balances: Dict[Address, U256] = field(default_factory=dict)
+    pre_nonces: Dict[Address, U64] = field(default_factory=dict)
```
Never used.
```suggestion
```

fselmo replied:
Yep this was left over, ty for catching!

spencer-tb commented on src/ethereum/forks/amsterdam/state_tracker.py:
```diff
@@ -0,0 +1,560 @@
+        assert (addr, key) in tx_frame.pre_storage
+        pre_value = tx_frame.pre_storage[(addr, key)]
+        post_value = tx_frame.storage_writes[(addr, key, idx)]
+        if (addr, key) in tx_frame.pre_storage:
```
If here is redundant following the assert.
```suggestion
```

spencer-tb commented on packages/testing/src/execution_testing/test_types/block_access_list/modifiers.py:
```diff
@@ -216,11 +227,16 @@ def modify_storage(
+        "code_changes",
+        BalCodeChange,
+        code,
+        "post_code",
```
`BalCodeChange` expects `"new_code"` parameter not `"post_code"`.
```suggestion
        "new_code",
```

spencer-tb commented on tests/amsterdam/eip7928_block_level_access_lists/test_cases.md:
```diff
@@ -21,16 +29,14 @@
-| `test_bal_system_contracts_2935_4788`            | BAL includes pre-exec system writes for parent hash & beacon root           | Build a block with `N` normal txs; 2935 & 4788 active                                                                          | BAL MUST include `HISTORY_STORAGE_ADDRESS` (EIP-2935) and `BEACON_ROOTS_ADDRESS` (EIP-4788) with `storage_changes` to ring-buffer slots; ea ...
-| `test_bal_system_dequeue_withdrawals_eip7002`    | BAL tracks post-exec system dequeues for withdrawals                        | Pre-populate EIP-7002 withdrawal requests; produce a block where dequeues occur                                                | BAL MUST include the 7002 system contract with `storage_changes` (queue head/tail slots 0–3) using `tx_index = len(txs)` and balance change ...
-| `test_bal_system_dequeue_consolidations_eip7251` | BAL tracks post-exec system dequeues for consolidations                     | Pre-populate EIP-7251 consolidation requests; produce a block where dequeues occur                                             | BAL MUST include the 7251 system contract with `storage_changes` (queue slots 0–3) using `tx_index = len(txs)`.                             ...
+| `test_bal_system_dequeue_consolidations_eip7251` | BAL tracks post-exec system dequeues for consolidations                     | Pre-populate EIP-7251 consolidation requests; produce a block where dequeues occur                                             | BAL MUST include the 7251 system contract with `storage_changes` (queue slots 0–3) using `block_access_index = len(txs)`.                   ...
```
Are these still planned? Maybe worth creating an issue for all remaining "Planned" BAL tests if so.

fselmo replied:
I added a line to the tracking issue [here](https://github.com/ethereum/execution-specs/issues/1878):

> All tests marked planned in [test_cases.md](https://github.com/ethereum/execution-specs/blob/forks/amsterdam/tests/amsterdam/eip7928_block_level_access_lists/test_cases.md) have been addressed / implemented.

spencer-tb commented on src/ethereum/forks/amsterdam/fork.py:
Do we need to check this? Or should we remove the TODO now?

fselmo replied:
This is in every fork in the specs and unrelated, but worth asking the question anyhow. Out of scope of this PR so going to mark resolved here. We can maybe address this in Discord and if we want to update this, we should PR the change to `forks/amsterdam` across all forks.

spencer-tb reviewed (approved):
Wanted to jump in on a review here. Thanks for pushing on with this. LGTM overall.
Just some questions and small comments!

fselmo committed:
chore: clean up based on comments on PR #1719

- Add CHANGELOG entry

fselmo commented:
@spencer-tb thanks for the review. I rebased and force pushed to get the coverage to be more accurate but only one commit added since your review (https://github.com/ethereum/execution-specs/pull/1719/commits/5b30479b74e2c612bb9146b2d74594468cd76376). Cleaned up 👍🏼.

We are just waiting on #2022 to be merged here and we can merge this to `forks/amsterdam`.

danceratopz commented on src/ethereum/forks/amsterdam/blocks.py:
```diff
@@ -242,6 +242,16 @@ class Header:
+    details.
+    [EIP-7928]: https://eips.ethereum.org/EIPS/eip-7928
+    [cbalh]: ref:ethereum.forks.amsterdam.block_access_lists.rlp_utils.compute_block_access_list_hash  # noqa: E501
+    """
```
We hash the block level access list using KECCAK256, but the docstring states SHA2-256. 

https://github.com/ethereum/execution-specs/blob/f3e3528923b602493547bfbcb005bfe2d784308b/src/ethereum/forks/amsterdam/block_access_lists/rlp_utils.py#L47-L48

```suggestion
    """
    [`keccak256`] hash of the Block Access List containing all accounts and
    storage locations accessed during block execution. Introduced in
    [EIP-7928]. See [`compute_block_access_list_hash`][cbalh] for more
    details.
    
    [`keccak256`]: ref:ethereum.crypto.hash.keccak256
    [EIP-7928]: https://eips.ethereum.org/EIPS/eip-7928
    [cbalh]: ref:ethereum.forks.amsterdam.block_access_lists.rlp_utils.compute_block_access_list_hash  # noqa: E501
    """
```

fselmo replied:
Nice catch! Ty! Squashed this into the last commit for PR comment cleanup 👍🏼

danceratopz commented on src/ethereum/forks/amsterdam/block_access_lists/rlp_utils.py:
```diff
@@ -0,0 +1,230 @@
+    return Bytes(encoded)
+
+
+def validate_block_access_list_against_execution(
```
`radon` flags this function with a cyclomatic complexity of 29 and [grades it D](https://radon.readthedocs.io/en/latest/commandline.html#the-cc-command) :) 

Tbh, I think it's very well laid out and readable with the comments. I just wanted to ask (@gurukamath, @fselmo,...) what the pattern is with complicated functions in EELS? Is there a precedent to break them down?

The next level complexity is 22 (which was pre-existing). Here are all `radon` warnings (CC>15) in forks/amsterdam:

| CC | Grade | Function | File | Line | Introduced by PR |
|----|-------|----------|------|------|------------------|
| 29 | D | `validate_block_access_list_against_execution` | `block_access_lists/rlp_utils.py` | 121 | EIP-7928 (this PR) |
| 22 | D | `check_transaction` | `fork.py` | 417 | Pre-existing |
| 18 | C | `recover_sender` | `transactions.py` | 650 | Pre-existing |
| 16 | C | `filter_net_zero_frame_changes` | `state_tracker.py` | 501 | EIP-7928 (this PR) |

As I say, it's readable as-is. Adding helper functions could actually make it harder to read.

Call and output:
```
uvx radon cc src/ethereum/forks/amsterdam/block_access_lists/rlp_utils.py -s
src/ethereum/forks/amsterdam/block_access_lists/rlp_utils.py
    F 121:0 validate_block_access_list_against_execution - D (29)
    F 51:0 rlp_encode_block_access_list - B (7)
    F 28:0 compute_block_access_list_hash - A (1)
```

fselmo replied:
I've asked about this before as this method isn't used anywhere in the current state of the specs. @nerolation do you think this is worth keeping around for the spec or was this more for a client implementation of a in-case-we-need-it scenario when you first implemented?

I would be ok removing this if it's not needed 👀

fselmo replied:
Okay, I removed this since it wasn't being used and squashed it into the single cleanup commit here. We can always add it back later if we need it.

danceratopz reviewed (approved):
Amazing collaboration by all participants on this PR! 

By no means a full review, just two random minor comments! 😄 Perhaps we could address the docstring in this PR. The complexity comment is more of an observation, respectively, a genuine question out of curiosity 🙂

gurukamath committed:
tests(eip-7928): generalize eip-7934 tests (#2022)

* tests(eip-7928): generalize eip-7934 tests

* ci(eip-7928): run all amsterdam tests on py3 and pypy3

---------

Co-authored-by: fselmo <fselmo2@gmail.com>
